src/libstore/lock.cc

```diff
@@ -7,6 +7,34 @@
 
 namespace nix {
 
+#if __linux__
+
+static std::vector<gid_t> get_group_list(const char *username, gid_t group_id)
+{
+    int ngroups {32}; // arbitrary initial guess
+    std::vector<gid_t> gids;
+    gids.resize(ngroups);
+
+    auto getgroupl {[&] {
+        return getgrouplist(username, group_id, gids.data(), &ngroups);
+    }};
+
+    if (-1 == getgroupl()) {
+        /* Our initial size of 32 wasn't sufficient, the
+           correct size has been stored in ngroups, so we try
+           again. */
+        gids.resize(ngroups);
+        if (-1 == getgroupl()) {
+            throw Error("failed to get list of supplementary groups for '%s'", username);
+        }
+    }
+
+    gids.resize(ngroups);
```

Comment:
With my above suggestion these go away; we've included the post-condition of resizing the vector to match the number of groups in the local function, so there is no need to do it separately.
```suggestion
        if (-1 == getgroupl()) {
            throw Error("failed to get list of supplementary groups for '%s'", username);
        }
    }

```